unison-cli/src/Unison/Codebase/Editor/Propagate.hs

```diff
@@ -265,10 +268,21 @@ propagate patch b = case validatePatch patch of
 
     Cli.Env {codebase} <- ask
     initialDirty <-
-      computeDirty
-        (liftIO . Codebase.dependents codebase Queries.ExcludeOwnComponent)
-        patch
-        (Names.contains names0)
+      liftIO $ Codebase.runTransaction codebase do
+        computeDirty
+          ( \ref -> do
+              mref <- case Cv.reference1to2 ref of
+                C.ReferenceBuiltin {} -> pure (Just ref)
+                C.ReferenceDerived id_ ->
+                  Operations.objectExistsForHash (view C.idH id_) <&> \case
+                    True -> Just ref
+                    False -> Nothing
+              case mref of
+                Nothing -> pure mempty
+                Just ref -> Codebase.dependents codebase Queries.ExcludeOwnComponent ref
+          )
+          patch
+          (Names.contains names0)
```

Comment:
It looks like this is the crux of the fix; but is this where it should go?  I think I convinced myself that the problem is more here than in [`computeDirty`](https://github.com/unisonweb/unison/blob/travis/3523/unison-cli/src/Unison/Codebase/Editor/Propagate.hs#L705-L726), but maybe it really belongs in [`Codebase.dependents`](https://github.com/unisonweb/unison/blob/travis/3523/codebase2/codebase-sqlite/U/Codebase/Sqlite/Operations.hs#L1035-L1040) instead; wdyt?

Reply:
We thought that it wouldn't be better to just relax a precondition on Codebase.dependents (that the given Reference must exist) without auditing the call sites. But maybe there aren't that many?

Reply:
Hrmm, semantically I think it makes sense that if we don't have a Reference in the codebase then the dependents of that reference must just be the empty set, are there any cases where you wouldn't expect that to be the case?

Reply:
I don't think so - the caller might expect a reference to be there, and we'd be hiding a bug by returning the empty set of dependents. I think there are simply two versions of this function with different semantics.

Reply:
It would be okay to audit the call sites; I don't know how many there are, but it doesn't seem obviously necessary either. (I imagine they would all be in the context of "how much work do we have to do about this reference" and if the reference is unknown, then there's no work to do.)

It would also be okay to just return the empty set when the query is an unknown reference.  The contract of `Codebase.dependents` is to return whatever dependents are present in this codebase. So if the object itself isn't in the codebase, there would also necessarily be zero dependents present. (There will always be some dependents missing from the result; the ones that don't exist in this codebase.)

`Codebase.dependencies` (if it existed, which it doesn't currently) would be a different story. It should always return the complete set of dependencies for any known object, and its return type would reflect that.

WDYT @mitchellwrosen

Reply:
Ah, ok. If we say the contract of Codebase.dependents is that it does not require the input reference to exist in the codebase. then of course it's the right place for the change. There weren't any docs on Codebase.dependents so we proceeded conservatively.

Reply:
I guess we have flexibility here, but the contract probably shouldn't be that it crashes.

Reply:
Let's go ahead and move it and maybe it will fix other things too :) /cc @tstat